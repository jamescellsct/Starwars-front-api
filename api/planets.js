import express from "express";

const router = express.Router();

var planets = require("./fakeData/db.json").planets;

router.get("/planets", function(req, res) {
  return res.status(200).json({
    count: planets.count,
    results: planets.results[0]
  });
});

router.get("/planets/:page", function(req, res) {
  let { page } = req.params;

  page = page.slice(5, 6) - 1;

  if (!planets.results[page]) {
    return res.status(404).send({
      error: {
        message: "Page Not Found!"
      }
    });
  }

  return res.status(200).json({
    count: planets.count,
    results: planets.results[page]
  });
});

export default router;
