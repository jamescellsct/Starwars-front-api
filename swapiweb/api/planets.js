import express from "express";

import { handleResulstWithId } from "./utils/handleResultsWithId";

const router = express.Router();

var planets = require("./fakeData/db.json").planets;

router.get("/planets", function(req, res) {
  return res.status(200).json({
    pageCount: planets.pageCount,
    results: handleResulstWithId(planets.pages[0], 1),
    page: 1
  });
});

router.get("/planets/:page", function(req, res) {
  let { page } = req.params;

  page = page.slice(5, 6) - 1;

  if (!planets.pages[page]) {
    return res.status(404).send({
      error: {
        message: "Page Not Found!"
      }
    });
  }

  return res.status(200).json({
    pageCount: planets.pageCount,
    results: handleResulstWithId(planets.pages[page], page + 1),
    page: page + 1
  });
});

export default router;
