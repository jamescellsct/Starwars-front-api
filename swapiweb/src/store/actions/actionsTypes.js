//Peoples Actions Types
const FETCH_PEOPLE_STARTED = "FETCH_PEOPLE_STARTED";
const FETCH_PEOPLE_SUCCEEDED = "FETCH_PEOPLE_SUCCEEDED";
const FETCH_PEOPLE_FAILED = "FETCH_PEOPLE_FAILED";
const UPDATE_PEOPLE_IS_FETCHING = "UPDATE_PEOPLE_IS_FETCHING";
const UPDATE_PEOPLE_SUCCEEDED = "UPDATE_PEOPLE_SUCCEEDED";
const UPDATE_PEOPLE_FAILED = "UPDATE_PEOPLE_FAILED";
export const peoplesAction = {
  FETCH_PEOPLE_STARTED,
  FETCH_PEOPLE_SUCCEEDED,
  FETCH_PEOPLE_FAILED,
  UPDATE_PEOPLE_IS_FETCHING,
  UPDATE_PEOPLE_SUCCEEDED,
  UPDATE_PEOPLE_FAILED
};

//Planets Actions Types
const FETCH_PLANETS_STARTED = "FETCH_PLANETS_STARTED";
const FETCH_PLANETS_SUCCEEDED = "FETCH_PLANETS_SUCCEEDED";
const FETCH_PLANETS_FAILED = "FETCH_PLANETS_FAILED";
const UPDATE_PLANETS_IS_FETCHING = "UPDATE_PLANETS_IS_FETCHING";
const UPDATE_PLANETS_SUCCEEDED = "UPDATE_PLANETS_SUCCEEDED";
const UPDATE_PLANETS_FAILED = "UPDATE_PLANETS_FAILED";
export const planetsAction = {
  FETCH_PLANETS_STARTED,
  FETCH_PLANETS_SUCCEEDED,
  FETCH_PLANETS_FAILED,
  UPDATE_PLANETS_IS_FETCHING,
  UPDATE_PLANETS_SUCCEEDED,
  UPDATE_PLANETS_FAILED
};

//Starships Actions Types
const FETCH_STARSHIPS_STARTED = "FETCH_STARSHIPS_STARTED";
const FETCH_STARSHIPS_SUCCEEDED = "FETCH_STARSHIPS_SUCCEEDED";
const FETCH_STARSHIPS_FAILED = "FETCH_STARSHIPS_FAILED";
const UPDATE_STARSHIPS_IS_FETCHING = "UPDATE_STARSHIPS_IS_FETCHING";
const UPDATE_STARSHIPS_SUCCEEDED = "UPDATE_STARSHIPS_SUCCEEDED";
const UPDATE_STARSHIPS_FAILED = "UPDATE_STARSHIPS_FAILED";
export const starshipsAction = {
  FETCH_STARSHIPS_STARTED,
  FETCH_STARSHIPS_SUCCEEDED,
  FETCH_STARSHIPS_FAILED,
  UPDATE_STARSHIPS_IS_FETCHING,
  UPDATE_STARSHIPS_SUCCEEDED,
  UPDATE_STARSHIPS_FAILED
};

//Vehicles Actions Types
const FETCH_VEHICLES_STARTED = "FETCH_VEHICLES_STARTED";
const FETCH_VEHICLES_SUCCEEDED = "FETCH_VEHICLES_SUCCEEDED";
const FETCH_VEHICLES_FAILED = "FETCH_VEHICLES_FAILED";
const UPDATE_VEHICLES_IS_FETCHING = "UPDATE_VEHICLES_IS_FETCHING";
const UPDATE_VEHICLES_SUCCEEDED = "UPDATE_VEHICLES_SUCCEEDED";
const UPDATE_VEHICLES_FAILED = "UPDATE_VEHICLES_FAILED";
export const vehiclesAction = {
  FETCH_VEHICLES_STARTED,
  FETCH_VEHICLES_SUCCEEDED,
  FETCH_VEHICLES_FAILED,
  UPDATE_VEHICLES_IS_FETCHING,
  UPDATE_VEHICLES_SUCCEEDED,
  UPDATE_VEHICLES_FAILED
};

//Pagination Action Types
const PAGE_SELECTED = "PAGE_SELECTED";
export const paginationAction = {
  PAGE_SELECTED
};
