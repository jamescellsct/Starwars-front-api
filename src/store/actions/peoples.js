import { peoplesAction } from "./actionsTypes";
import { getPeople } from "../../services/api/peopleService";
import { handleResulstWithId } from "../../utils/handleResultsWithId";

const fetchPeopleStarted = () => ({
  type: peoplesAction.FETCH_PEOPLE_STARTED
});

const fetchPeopleSucceeded = people => ({
  type: peoplesAction.FETCH_PEOPLE_SUCCEEDED,
  people
});

const fetchPeopleFailed = error => ({
  type: peoplesAction.FETCH_PEOPLE_FAILED,
  error
});

export const fetchPeople = () => dispatch => {
  dispatch(fetchPeopleStarted());

  return getPeople()
    .then(response => {
      const { results, count } = response;
      dispatch(
        fetchPeopleSucceeded({
          page: 1,
          results: handleResulstWithId(results, 1),
          pageCount: Math.ceil(count / 10)
        })
      );
    })
    .catch(error => {
      dispatch(fetchPeopleFailed(error));
    });
};

const updatePeopleIsFetching = () => ({
  type: peoplesAction.UPDATE_PEOPLE_IS_FETCHING
});

const updatePeopleSucceeded = people => ({
  type: peoplesAction.UPDATE_PEOPLE_SUCCEEDED,
  people
});

const updatePeopleFailed = error => ({
  type: peoplesAction.UPDATE_PEOPLE_FAILED,
  error
});

export const updatePeople = page => dispatch => {
  dispatch(updatePeopleIsFetching());

  getPeople(page)
    .then(response => {
      const { results, count } = response;
      dispatch(
        updatePeopleSucceeded({
          page,
          results: handleResulstWithId(results, page),
          pageCount: Math.ceil(count / 10)
        })
      );
    })
    .catch(error => {
      dispatch(updatePeopleFailed(error));
    });
};
