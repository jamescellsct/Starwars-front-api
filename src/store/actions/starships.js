import { starshipsAction } from "./actionsTypes";
import { getStarships } from "../../services/api/starshipsService";
import { handleResulstWithId } from "../../utils/handleResultsWithId";

const fetchStarshipsStarted = () => ({
  type: starshipsAction.FETCH_STARSHIPS_STARTED
});

const fetchStarshipsSucceeded = starships => ({
  type: starshipsAction.FETCH_STARSHIPS_SUCCEEDED,
  starships
});

const fetchStarshipsFailed = error => ({
  type: starshipsAction.FETCH_STARSHIPS_FAILED,
  error
});

export const fetchStarships = () => dispatch => {
  dispatch(fetchStarshipsStarted());

  return getStarships()
    .then(response => {
      const { results, count } = response;
      dispatch(
        fetchStarshipsSucceeded({
          page: 1,
          results: handleResulstWithId(results, 1),
          pageCount: Math.ceil(count / 10)
        })
      );
    })
    .catch(error => {
      dispatch(fetchStarshipsFailed(error));
    });
};

const updateStarshipsIsFetching = () => ({
  type: starshipsAction.UPDATE_STARSHIPS_IS_FETCHING
});

const updateStarshipsSucceeded = starships => ({
  type: starshipsAction.UPDATE_STARSHIPS_SUCCEEDED,
  starships
});

const updateStarshipsFailed = error => ({
  type: starshipsAction.UPDATE_STARSHIPS_FAILED,
  error
});

export const updateStarships = page => dispatch => {
  dispatch(updateStarshipsIsFetching());
  return getStarships(page)
    .then(response => {
      const { results, count } = response;
      dispatch(
        updateStarshipsSucceeded({
          page,
          results: handleResulstWithId(results, page),
          pageCount: Math.ceil(count / 10)
        })
      );
    })
    .catch(error => {
      dispatch(updateStarshipsFailed(error));
    });
};
