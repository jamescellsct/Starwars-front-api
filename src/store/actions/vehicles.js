import { vehiclesAction } from "./actionsTypes";
import { getVehicles } from "../../services/api/vehiclesService";
import { handleResulstWithId } from "../../utils/handleResultsWithId";

const fetchVehiclesStarted = () => ({
  type: vehiclesAction.FETCH_VEHICLES_STARTED
});

const fetchVehiclesSucceeded = vehicles => ({
  type: vehiclesAction.FETCH_VEHICLES_SUCCEEDED,
  vehicles
});

const fetchVehiclesFailed = error => ({
  type: vehiclesAction.FETCH_VEHICLES_FAILED,
  error
});

export const fetchVehicles = () => dispatch => {
  dispatch(fetchVehiclesStarted());

  return getVehicles()
    .then(response => {
      const { results, count } = response;
      dispatch(
        fetchVehiclesSucceeded({
          page: 1,
          results: handleResulstWithId(results, 1),
          pageCount: Math.ceil(count / 10)
        })
      );
    })
    .catch(error => {
      dispatch(fetchVehiclesFailed(error));
    });
};

const updateVehiclesIsFetching = () => ({
  type: vehiclesAction.UPDATE_VEHICLES_IS_FETCHING
});

const updateVehiclesSucceeded = vehicles => ({
  type: vehiclesAction.UPDATE_VEHICLES_SUCCEEDED,
  vehicles
});

const updateVehiclesFailed = error => ({
  type: vehiclesAction.UPDATE_VEHICLES_FAILED,
  error
});

export const updateVehicles = page => dispatch => {
  dispatch(updateVehiclesIsFetching());
  return getVehicles(page)
    .then(response => {
      const { results, count } = response;
      dispatch(
        updateVehiclesSucceeded({
          page,
          results: handleResulstWithId(results, page),
          pageCount: Math.ceil(count / 10)
        })
      );
    })
    .catch(error => {
      dispatch(updateVehiclesFailed(error));
    });
};
