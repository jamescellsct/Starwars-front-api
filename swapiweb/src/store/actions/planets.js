import { planetsAction } from "./actionsTypes";
import { getPlanets } from "../../services/api/planetsService";
import { handleResulstWithId } from "../../utils/handleResultsWithId";

const fetchPlanetsStarted = () => ({
  type: planetsAction.FETCH_PLANETS_STARTED
});

const fetchPlanetsSucceeded = planets => ({
  type: planetsAction.FETCH_PLANETS_SUCCEEDED,
  planets
});

const fetchPlanetsFailed = error => ({
  type: planetsAction.FETCH_PLANETS_FAILED,
  error
});

export const fetchPlanets = () => dispatch => {
  dispatch(fetchPlanetsStarted());

  getPlanets()
    .then(response => {
      const { results, count } = response;
      dispatch(
        fetchPlanetsSucceeded({
          page: 1,
          results: handleResulstWithId(results, 1),
          pageCount: Math.ceil(count / 10)
        })
      );
    })
    .catch(error => {
      dispatch(fetchPlanetsFailed(error));
    });
};

const updatePlanetsIsFetching = () => ({
  type: planetsAction.UPDATE_PLANETS_IS_FETCHING
});

const updatePlanetsSucceeded = planets => ({
  type: planetsAction.UPDATE_PLANETS_SUCCEEDED,
  planets
});

const updatePlanetsFailed = error => ({
  type: planetsAction.UPDATE_PLANETS_FAILED,
  error
});

export const updatePlanets = page => dispatch => {
  dispatch(updatePlanetsIsFetching());
  getPlanets(page)
    .then(response => {
      const { results, count } = response;
      dispatch(
        updatePlanetsSucceeded({
          page,
          results: handleResulstWithId(results, page),
          pageCount: Math.ceil(count / 10)
        })
      );
    })
    .catch(error => {
      dispatch(updatePlanetsFailed(error));
    });
};
