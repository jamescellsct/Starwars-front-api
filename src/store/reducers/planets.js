import { planetsAction } from "../actions/actionsTypes";

export default function planets(state = null, action) {
  switch (action.type) {
    case planetsAction.FETCH_PLANETS_STARTED:
      return Object.assign({}, state, {
        isFetching: true
      });

    case planetsAction.FETCH_PLANETS_SUCCEEDED:
      return Object.assign({}, state, {
        pages: [
          ...(state.pages || []),
          { page: action.planets.page, list: action.planets.results }
        ],
        pageCount: action.planets.pageCount,
        isFetching: false
      });

    case planetsAction.FETCH_PLANETS_FAILED:
      return Object.assign({}, state, {
        isFetching: false,
        error: action.error
      });

    case planetsAction.UPDATE_PLANETS_IS_FETCHING:
      return Object.assign({}, state, {
        isFetching: true
      });
    case planetsAction.UPDATE_PLANETS_SUCCEEDED:
      return Object.assign({}, state, {
        pages: [
          ...state.pages,
          { page: action.planets.page, list: action.planets.results }
        ],
        pageCount: action.planets.pageCount,
        isFetching: false
      });

    case planetsAction.UPDATE_PLANETS_FAILED:
      return Object.assign({}, state, {
        isFetching: false,
        error: action.error
      });

    default:
      return state;
  }
}
