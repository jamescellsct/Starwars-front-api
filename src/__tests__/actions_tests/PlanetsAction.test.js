import thunk from "redux-thunk";
import configureMockStore from "redux-mock-store";
import nock from "nock";

import { planetsAction } from "../../store/actions/actionsTypes";
import { fetchPlanets, updatePlanets } from "../../store/actions/planets";
import BASE_URL from "./baseUrl";

const midlewares = [thunk];
const mockStore = configureMockStore(midlewares);

const mockResults = [
  {
    name: "Alma Hopkins",
    avatar: "http://i.pravatar.cc/40?img=0.28152803167206786",
    rotation_period: 128,
    orbital_period: 459,
    diameter: 7065,
    climate: "murky",
    gravity: 2.5,
    terrain: "swamp",
    surface_water: 37,
    population: 38727741060,
    idItem: 1
  }
];

const mockUpdateResults = [
  {
    name: "Person Updating",
    avatar: "http://i.pravatar.cc/40?img=0.28152803167206786",
    rotation_period: 128,
    orbital_period: 459,
    diameter: 7065,
    climate: "murky",
    gravity: 2.5,
    terrain: "swamp",
    surface_water: 37,
    population: 38727741060,
    idItem: 1
  }
];

describe("Testing Planets Action", () => {
  it("Shold create an action to fetching planets data", () => {
    afterEach(() => {
      nock.cleanAll();
    });

    nock(BASE_URL)
      .defaultReplyHeaders({ "access-control-allow-origin": "*" })
      .get("/planets")
      .reply(200, {
        count: 20,
        results: mockResults
      });

    const expectedAction = [
      { type: planetsAction.FETCH_PLANETS_STARTED },
      {
        type: planetsAction.FETCH_PLANETS_SUCCEEDED,
        planets: {
          page: 1,
          results: mockResults,
          pageCount: 2
        }
      }
    ];

    const store = mockStore({ planets: {} });
    return store.dispatch(fetchPlanets()).then(() => {
      expect(store.getActions()).toEqual(expectedAction);
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it("Should return ERROR in fetching planets data", () => {
    afterEach(() => {
      nock.cleanAll();
    });

    nock(BASE_URL)
      .defaultReplyHeaders({ "access-control-allow-origin": "*" })
      .get("/planets")
      .reply(400, {});

    const expectedAction = [
      { type: planetsAction.FETCH_PLANETS_STARTED },
      {
        type: planetsAction.FETCH_PLANETS_FAILED,
        error: new Error("Request failed with status code 400")
      }
    ];

    const store = mockStore({ planets: {} });
    return store.dispatch(fetchPlanets()).then(() => {
      expect(store.getActions()).toEqual(expectedAction);
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it("Should create an action to update planets data", () => {
    afterEach(() => {
      nock.cleanAll();
    });

    nock(BASE_URL)
      .defaultReplyHeaders({ "access-control-allow-origin": "*" })
      .get("/planets/page=1")
      .reply(200, {
        count: 20,
        results: mockUpdateResults
      });

    const expectedAction = [
      { type: planetsAction.UPDATE_PLANETS_IS_FETCHING },
      {
        type: planetsAction.UPDATE_PLANETS_SUCCEEDED,
        planets: {
          page: 1,
          results: mockUpdateResults,
          pageCount: 2
        }
      }
    ];

    const store = mockStore({ planets: {} });
    return store.dispatch(updatePlanets(1)).then(() => {
      expect(store.getActions()).toEqual(expectedAction);
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it("Should return ERROR on updating planets data", () => {
    afterEach(() => {
      nock.cleanAll();
    });

    nock(BASE_URL)
      .defaultReplyHeaders({ "access-control-allow-origin": "*" })
      .get("/planets/page=1")
      .reply(400, {});

    const expectedAction = [
      { type: planetsAction.UPDATE_PLANETS_IS_FETCHING },
      {
        type: planetsAction.UPDATE_PLANETS_FAILED,
        error: new Error("Request failed with status code 400")
      }
    ];

    const store = mockStore({ planets: {} });
    return store.dispatch(updatePlanets(1)).then(() => {
      expect(store.getActions()).toEqual(expectedAction);
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
