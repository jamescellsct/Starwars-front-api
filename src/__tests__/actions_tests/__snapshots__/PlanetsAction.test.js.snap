// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Planets Action Shold create an action to fetching planets data 1`] = `
Array [
  Object {
    "type": "FETCH_PLANETS_STARTED",
  },
  Object {
    "planets": Object {
      "page": 1,
      "pageCount": 2,
      "results": Array [
        Object {
          "avatar": "http://i.pravatar.cc/40?img=0.28152803167206786",
          "climate": "murky",
          "diameter": 7065,
          "gravity": 2.5,
          "idItem": 1,
          "name": "Alma Hopkins",
          "orbital_period": 459,
          "population": 38727741060,
          "rotation_period": 128,
          "surface_water": 37,
          "terrain": "swamp",
        },
      ],
    },
    "type": "FETCH_PLANETS_SUCCEEDED",
  },
]
`;
